### PR TITLE
Clock App #32 : add countdown / timer functionality 

### PR DESCRIPTION
**Pull request by dark-eye**
Added a Timer page ( Issue #32 ) :
- Currently support only one active timer.
- Currently only have analog clock UI.
- Support for saving timers for later use.
- An active timer will not be shown in the alarm list as to not confuse the user.

Also : 
- Changed the startup animation to be a little fast and using standard UbuntuAnimation values.
- Made the rest of the pages load only after the animation ends to make the loading time feel a little faster.
- Fixed an issue with the fastscroll UI.
- Changed the navigation row shodow a little more "shadowy" (had hard edges before it`s now a little softer).
- Hide the clock fold in the stopwatch (it didn't actually folded when clicked)

### CODE REVIEW

**Comment by dark-eye** [comment]
Here is a click for armhf  of this version : (	
com.ubuntu.clock_3.9.560_armhf.click)[https://my.owndrive.com/index.php/s/YkMStcoAQ1e078s]

**Comment by Flohack74** [comment]
We got a lot of commits here, time to merge I would say...

File: app/timer/TimerPage.qml [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+            id: startStopButton
+
+            property bool inProgress: false
+
+            objectName: "startAndStopButton"
+            width: _timerPage.width / 2 - units.gu(1)
+            height: units.gu(4)
+            enabled: !inProgress  && (isRunning || timerFace.getCircle().hasTime)
+            color: !isRunning  ? UbuntuColors.green : UbuntuColors.red
+            text: isRunning ? i18n.tr("Stop") : (true ? i18n.tr("Start") : i18n.tr("Resume"))
```
Comment (sverzegnassi) [inline app/timer/TimerPage.qml]: `text: isRunning ? i18n.tr("Stop") : i18n.tr("Start")`

**Comment by dark-eye** [comment]
Hey @sverzegnassi 

I`m working on fixing the code as request on some of your comment however i have a few points about the changes i made an why i made them :


>    Revert X-Ubuntu-Splash-Color from #A0A0A0 to #F5F5F5

This was changed due to the fact that adding a dark theme to be used at night is completely negated by the fact that the clock start with a bright white splash color, might i suggest that if the gray (middle ground ) doesn't work we should go to full black like some other applications?

>   Fix alignment and spacing for controls. The timer face and { "Start" | "Stop" } buttons have a different spacing than other pages.

This is an issue as you need to have enough space for the digital time display at the top I can align the buttons, But the timer clock will have to be a little lower then the other clocks as to display the current selected digital time, I am planing to fix it i the future (by changing the UI  somehow) but i don't think an alignment issue should stop this feature.

>  Remove the vibration when dragging an hand in the timer face. On BQ devices, it's quite annoying (the E5 is a bit wonky).

This was added as a result that it felt more responsive and less wonky to feel the time changes in meizu Pro , in any case vibrations  removed...

>  Although my preference would go for the standard (i.e. non-circular) picker from Ubuntu UI Toolkit, I have at least to require you to use the hands visuals from the main "Clock" page, and improve the quality of the ticks on the timer face (they look blurry)

I use the analog clock as the UITK picker doesn't work very well and need some fixing (i have a branch where i started to work on it ), About the clock hands visual there are the same as the regular clock (in fact it uses the analog clock class as a base).
i`ll try and create some non-blurry tick marks :)

>   Allow users to set up timers that last less than one minute

This can't actually be done without fixing the core implementation of the alarm system as i think it`s cron based and might fire immediately after setting the alarm or after some time unrelated to actual amount of time left  (behavior seems random...) 

>   The document-save icon is ambiguous in this context. We might consider to create ad-hoc icon

I really don't want to use none standard icons in core app might i suggest the  `like` or `star` icons?

>   At least in my case, I found impossible to save a timer. Keyboard and bottom navigation completely obscured the TextField. I can't either scroll the page. See screenshot.

Was kinda aware that might be an issue as in meizu pro you can exactly to the bottom of the text field, I'll fix it by moving the text field to the same line as the  start button and i`ll try to  hide the bottom navigation on virtual keyboard showing.

**Comment by dark-eye** [comment]
Still Working on it just uploaded the current commits so i can work it tomorrow from another computer

**Comment by sverzegnassi** [comment]
Replying point by point:

> This was changed due to the fact that adding a dark theme to be used at night is completely negated by the fact that the clock start with a bright white splash color, might i suggest that if the gray (middle ground ) doesn't work we should go to full black like some other applications?

In that case, I'd suggest to use colors from the official palette. `#111111` (a.k.a. `UbuntuColors.Jet`) is one of them, and it's already used for the indicators bar in Unity 8. That way, you get a visual continuity between the different components of Unity.

In any case, please postpone this bugfix until night mode will be introduced by a separate PR, so that we can discuss about it more linearly and we have the full feature in front of our eyes :)

> This can't actually be done without fixing the core implementation of the alarm system as i think it`s cron based and might fire immediately after setting the alarm or after some time unrelated to actual amount of time left (behavior seems random...)

Let's address it once more stable Xenial images are available. We might want to have some backend in the system for this (like it has been done for alarms).

> I really don't want to use none standard icons in core app might i suggest the like or star icons?

Have a check at: http://icons.mivoligo.com/
I'm not sure how it would look like, but `add-to-playlist` or any of those `*-new` might adapt better to the different context.

> Was kinda aware that might be an issue as in meizu pro you can exactly to the bottom of the text field, I'll fix it by moving the text field to the same line as the start button and i`ll try to hide the bottom navigation on virtual keyboard showing.

Sure, no prob.

> This is an issue as you need to have enough space for the digital time display at the top I can align the buttons, But the timer clock will have to be a little lower then the other clocks as to display the current selected digital time, I am planing to fix it i the future (by changing the UI somehow) but i don't think an alignment issue should stop this feature.

One of the main characteristics that attracted much interest on Ubuntu Touch over time was the UX design. It has always been the focus for the entire project, and this is something we would like not to change.
For us, visual consistency and good UX are just as important as other *regular* features.

We appreciated when Mitu intervened on the bug tracker, for recalling this aspect. He has been active on Ubuntu Touch UX for the last 5 years, when the Ubuntu community was still using Balsamiq for the core apps mockups, so he's very aware of the importance of good visuals.

This is also the reason why we still call the OS "Ubuntu Touch". Canonical authorized the usage of their brand name not just as a simple remix; as a consequence, we need to put our best efforts in providing the best visuals when it comes to brand recognition.

Small breakage in UI might be less relevant, when the introduced feature is somehow important. However, I'm sorry to say that it might not be the case this time. 

This is how the Timer tab currently appears, compared with a regular tab of the old Canonical clock-app. Visual specs do not match anymore.

Current - screenshot

We need to - at least, try to - match the old specs, which are still valid despite the introduction of the bottom navigation.

Proposed - screenshot

In future, we might want to give you full write permissions on the repo, so that you don't need to wait for @Flohack74 or me to review your PR. :-)

We appreciate your contribution here, and we trust your work. But we need to be sure that design issues would receive the due attention, and can be addressed before publishing the app on the store

**Comment by dark-eye** [comment]
Hey @sverzegnassi ,

Wasn't aware about all the design specification and the issue my changes will cause i`ll try andd adhere to them better in the future, Do you know of better specification other then [what seems to be a old design](https://docs.google.com/presentation/d/1JvDyhsW17d1-Mz8OY1YMBKwfRI2z9qgyRjbujEsxEMk/edit#slide=id.g18895458d_00)?

Regarding the splash i already applied the night mode (or suru dark) theme in previous PR so will include the change in this PR.

I've aligned the design elements and changed the icon to a new alarm icon which fit the intention perfectly (thanks for the link :) ).

Thank you  for the trust, i`ll try my beast to help the project as best as i can ,  i will probably never merge my own code without at least one other person approving it but will be happy to review other people code :)

**Comment by sverzegnassi** [comment]
[Clock Visual Refresh.pdf](https://github.com/ubports/clock-app/files/1569295/Clock.Visual.Refresh.pdf)

**Comment by sverzegnassi** [comment]
Looks much better by far, great job! And the issue with adding new timer presets is gone now
Could you please confirm your work on this PR is completed, so that we can merge it?

**Comment by dark-eye** [comment]
@sverzegnassi As software does i just noticed an issue on saving new timers, investigating...

**Comment by dark-eye** [comment]
@sverzegnassi  Weird it seems that when entering the name of a timer to save if you don't press space or any none word character then the add new timer button stays disabled.
My guess is that this issue has to do with the keyboard auto-completed and TextField interaction i've  changed the watched property from text to displayText and added a visual aid to the button to indicate that it is disabled.

**Comment by sverzegnassi** [comment]
@dark-eye Yes, it's the auto-completion. You might want to check for preedit text and use [QInputMethod::commit()](http://doc.qt.io/qt-5/qinputmethod.html#commit)

IIRC, in QML it is:
```
if (Qt.inputMethod.preeditText) {
    Qt.inputMethod.commit()
}
// Save alarm
```

or you can disable auto-completion by setting `inputMethodHints :Qt.ImhNoPredictiveText` for your TextField

**Comment by dark-eye** [comment]
@sverzegnassi  Yes it good by me :)

**Comment by sverzegnassi** [comment]
Cool, thanks again. Great work! :)

**Pull request merged**